pkg/clients/firewall/firewall.go

```diff
@@ -0,0 +1,139 @@
+				Ports:      rule.Ports,
+			}
+		}
+	}
+
+	if in.LogConfig != nil {
+		firewall.LogConfig = &compute.FirewallLogConfig{
+			Enable: in.LogConfig.Enable,
+		}
+	}
+}
+
+// GenerateFirewallObservation takes a compute.Firewall and returns *FirewallObservation.
+func GenerateFirewallObservation(in compute.Firewall) v1alpha1.FirewallObservation {
+	fw := v1alpha1.FirewallObservation{
+		CreationTimestamp: in.CreationTimestamp,
+		ID:                in.Id,
+		SelfLink:          in.SelfLink,
+	}
+	return fw
+}
+
+// LateInitializeSpec fills unassigned fields with the values in compute.Firewall object.
+func LateInitializeSpec(spec *v1alpha1.FirewallParameters, in compute.Firewall) {
+	spec.Disabled = gcp.LateInitializeBool(spec.Disabled, in.Disabled)
+	spec.Network = gcp.LateInitializeString(spec.Network, in.Network)
+	spec.Priority = gcp.LateInitializeInt64(spec.Priority, in.Priority)
+	spec.Description = gcp.LateInitializeString(spec.Description, in.Description)
+	spec.Direction = gcp.LateInitializeString(spec.Direction, in.Direction)
+	spec.SourceRanges = gcp.LateInitializeStringSlice(spec.SourceRanges, in.SourceRanges)
+	spec.DestinationRanges = gcp.LateInitializeStringSlice(spec.DestinationRanges, in.DestinationRanges)
+	spec.SourceServiceAccounts = gcp.LateInitializeStringSlice(spec.SourceServiceAccounts, in.SourceServiceAccounts)
+	spec.TargetServiceAccounts = gcp.LateInitializeStringSlice(spec.TargetServiceAccounts, in.TargetServiceAccounts)
+	spec.SourceTags = gcp.LateInitializeStringSlice(spec.SourceTags, in.SourceTags)
+	spec.TargetTags = gcp.LateInitializeStringSlice(spec.TargetTags, in.TargetTags)
+
+	if in.LogConfig != nil && spec.LogConfig == nil {
+		spec.LogConfig = &v1alpha1.FirewallLogConfig{
+			Enable: in.LogConfig.Enable,
+		}
+	}
+
+	if len(in.Allowed) != 0 && len(spec.Allowed) == 0 {
+		spec.Allowed = make([]*v1alpha1.FirewallAllowed, len(in.Allowed))
+		for idx, rule := range in.Allowed {
+			spec.Allowed[idx] = &v1alpha1.FirewallAllowed{
+				IPProtocol: rule.IPProtocol,
+				Ports:      rule.Ports,
+			}
+		}
+	}
+
+	if len(in.Denied) != 0 && len(spec.Denied) == 0 {
+		spec.Denied = make([]*v1alpha1.FirewallDenied, len(in.Allowed))
+		for idx, rule := range in.Denied {
+			spec.Denied[idx] = &v1alpha1.FirewallDenied{
+				IPProtocol: rule.IPProtocol,
+				Ports:      rule.Ports,
+			}
+		}
+	}
+}
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(name string, in *v1alpha1.FirewallParameters, observed *compute.Firewall) (upTodate bool, err error) {
+	generated, err := copystructure.Copy(observed)
+	if err != nil {
+		return true, errors.Wrap(err, errCheckUpToDate)
+	}
+	desired, ok := generated.(*compute.Firewall)
+	if !ok {
+		return true, errors.New(errCheckUpToDate)
+	}
+	GenerateFirewall(name, *in, desired)
+	return cmp.Equal(desired, observed, cmpopts.EquateEmpty(), cmpopts.IgnoreFields(compute.Firewall{}, "ForceSendFields")), nil
```

Comment:
While testing on my side, I noticed `IsUpToDate` keeps returning `false`. To debug further, I've added following lines just above return:

```
	diff := cmp.Diff(desired, observed, cmpopts.EquateEmpty(), cmpopts.IgnoreFields(compute.Firewall{}, "ForceSendFields"))
	fmt.Println(diff)
```

Which printed:
```
  &compute.Firewall{
        ... // 9 identical fields
        LogConfig: &{},
        Name:      "firewall-example-by-hasan",
-       Network:   "projects/wesaas-playground/global/networks/example-by-hasan",
+       Network:   "https://www.googleapis.com/compute/v1/projects/wesaas-playground/global/networks/example-by-hasan",
        Priority:  1000,
        SelfLink:  "https://www.googleapis.com/compute/v1/projects/wesaas-playground"...,
        ... // 1 ignored and 7 identical fields
  }
```

According [GCP API reference](https://cloud.google.com/compute/docs/reference/rest/v1/firewalls), both are valid for network field, however, we count this as a diff and keep sending updates to the resource.